opentimsj/src/main/native/timsj_TimsTOFExperimentReader.cpp

```diff
@@ -7,8 +6,86 @@
 #include <cstdlib>
 #include <stdio.h>
 
+#include "timsj_TimsTOFExperimentReader.h"
 #include "../../../../opentims++/opentims_all.cpp"
 
+
+struct TimsFramePl {
+    int frameId; // coordinate
+    std::vector<double> mzs, intensities, inv_ion_mobs; // data
+    std::vector<int> scans, tofs; // data
+    // constructors
+    TimsFramePl(){}
+    TimsFramePl(int id, std::vector<int> scan, std::vector<double> mz, std::vector<double> intensity, std::vector<int> tof, std::vector<double> inv_ion_mob):
+    frameId(id), mzs(mz), intensities(intensity), scans(scan), tofs(tof), inv_ion_mobs(inv_ion_mob){}
+};
+
+/**
+ * helper for TDH initialization
+ * @param dp data path
+ * @param bp binary path
+ * @return a tims data handle
+ */
+TimsDataHandle get_tdh(std::string dp, std::string bp){
+    // need to translate TOF to MZ
+    DefaultTof2MzConverterFactory::setAsDefault<BrukerTof2MzConverterFactory, const char*>(bp.c_str());
+    // need to translate SCAN to 1/K0
+    DefaultScan2InvIonMobilityConverterFactory::setAsDefault<BrukerScan2InvIonMobilityConverterFactory, const char*>(bp.c_str());
+
+    return TimsDataHandle(dp);
+}
+
+/**
+ *
+ */
+class ExposedTimsDataHandle{
+private:
+    TimsDataHandle handle;
+
+public:
+    // path to dataset, path to bruker binaries
+    std::string datasetPath, binaryPath;
+    // handle to read from raw data
+    ExposedTimsDataHandle(std::string dp, std::string bp);
+
+    TimsFramePl getTimsFramePl(const int frameId);
+};
+
+ExposedTimsDataHandle::ExposedTimsDataHandle(std::string dp, std::string bp) : handle(get_tdh(dp, bp)) {
+    datasetPath = dp;
+    binaryPath = bp;
+};
+
+
+TimsFramePl ExposedTimsDataHandle::getTimsFramePl(const int frameId) {
+    // allocate buffer
+    const size_t buffer_size_needed = handle.max_peaks_in_frame();
+    std::unique_ptr<uint32_t[]> scan_ids = std::make_unique<uint32_t[]>(buffer_size_needed);
+    std::unique_ptr<uint32_t[]> intens = std::make_unique<uint32_t[]>(buffer_size_needed);
+    std::unique_ptr<uint32_t[]> tofs = std::make_unique<uint32_t[]>(buffer_size_needed);
+    std::unique_ptr<double[]> inv_ion_mob = std::make_unique<double[]>(buffer_size_needed);
+    std::unique_ptr<double[]> mz = std::make_unique<double[]>(buffer_size_needed);
+
+    // fetch
+    TimsFrame& frame = handle.get_frame(frameId);
+    frame.save_to_buffs(nullptr, scan_ids.get(), tofs.get(), intens.get(), mz.get(), inv_ion_mob.get(), nullptr);
+
+    // allocate concrete vectors
+    std::vector<double> mzs, intensities, inv_ion_mobility;
+    std::vector<int> scans, tof;
+
```

Comment:
You might want to do a .reserve(frame.num_peaks) here on each of those vectors for more speed gains ;)

Reply:
good point added this to all vectors that get filled on this call.